### ISSUE TITLE
Submission API

### ISSUE THREAD

**Issue by sreejithr** [GoChallenge/gochallenge#6]
Discussion forked off of #4 following [suggestion](https://github.com/GoChallenge/gochallenge/issues/4#issuecomment-78438217).

Initial ideas by @gedex for this endpoint involved embedding the plaintext of files inside the request's json itself. This is fine for small files. To handle larger files, a multipart transmission seems fit.

`gochallenge submit` should ask the user to confirm that he's submitting to avoid accidental submissions. Following confirmation, we should tarball/zip the current folder and proceed to `POST`.

@IndianGuru expressed [concern](https://github.com/GoChallenge/gochallenge/issues/4#issuecomment-78440058) about resubmissions. If resubmissions make the evaluation process go haywire, I feel we shouldn't support it (for the time being?) . Any further submission should be blocked.

But I think the **user should be able to delete his submission at any point** since the content belongs to him. `gochallenge rm` following a confirmation would do. Further submits after a `rm` should be prevented.
## Create Submission

```
POST /v1/challenges/:challenge_id/submission/
```

```
Content-type: multipart/related
```
### Input

Say, `--foo_bar_baz` being the boundary

```
--foo_bar_baz
Content-Type: application/json; charset=UTF-8

{
    "file_name": "challenge-1.tar.gz"
}

--foo_bar_baz
Content-Type: application/octet-stream

tar data
--foo_bar_baz--
```
### Response

```
{
     "id": 3,
     "checksum": <checksum>,
     "user": {
          ....
     },
     "created": "2010-04-14T02:15:15Z",
     "modified": "2010-04-14T02:15:15Z"
}
```

The complexities of multipart upload is anyway abstracted out by the multipart package in Go. We will return the checksum (md5?) for the file for the user to verify.
### Status

```
201 OK
```
## Delete submission

```
DELETE /v1/challenges/:challenge_id/submission/
```
### Response

```
{
    "success": true,
    "deleted": "2010-04-14T02:15:15Z"
}
```
### Status

```
200 OK
```

Post your feedback.

P.S: Not talking about sanity check of the files before submission since the topic warrants an issue of its own.

**Comment by morhekil** [comment]
just a few quick notes, to sum up what was discussed so far.

First off, there's an early draft of this data available on wiki [here](https://github.com/GoChallenge/gochallenge/wiki/API-v1#create-a-submission) - it adds a hash of the data for verification, and type of the indicate user participation.

I do like the idea of multipart upload, dumping a big blog inside json felt weird to me - multipart feels like a much cleaner solution.

> Any further submission should be blocked.

this is probably a server-side concern, we can't trust client-side tool to implement this check securely. Server will return an error, cli should just display it.

> user should be able to delete submission at any point

I disagree. In my view, but submitting a solution user effectively surrenders further control of the code to the challenge organisers. Think of it like "whets gets on the internet - stays on the internet". Participation is voluntary, and no one is forced to send the submission, but when one is send - it results in certain work of other people (evaluators, organisers), which should be honoured just as well. 

If we decide to go ahead with a cooling-off period for submissions, we might allow to remove a submission during the same period, but once it's up for evaluation - it's locked.

**Comment by morhekil** [comment]
I updated the spec page of the wiki with your suggestions about create submission part, thanks for that! Feel free to update it if I missed/misinterpreted anything

**Comment by sreejithr** [comment]
Looks good. Has somebody started working on the API?

**Comment by morhekil** [comment]
yes, last night I build the basics of submission-related code, you can find the current version in [apiv1 branch](https://github.com/GoChallenge/gochallenge/tree/apiv1). It implements your idea of multipart submission, and cli tools also has support for it already. [Here](https://asciinema.org/a/17721) is the asciicast of the current workflow.

Feel free to jump in if you want to help with some code - there's still a lot of work left on the server, from proper error reporting to data persistence and a frontend.
